src/toxcore/dht_new/packet.rs

```diff
@@ -0,0 +1,938 @@
+
+It is used in many places in Tox, e.g. in `DHT Send nodes`.
+
+To store more than one node, simply append another on to the previous one:
+
+`[packed node 1][packed node 2][...]`
+
+Serialized Packed node:
+
+Length | Content
+------ | -------
+1      | [`IpType`](./.enum.IpType.html)
+4 or 16| IPv4 or IPv6 address
+2      | port
+32     | node ID
+
+Size of serialized `PackedNode` is 39 bytes with IPv4 node info, or 51 with
+IPv6 node info.
+
+DHT module *should* use only UDP variants of `IpType`, given that DHT runs
+solely on the UDP.
+*/
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct PackedNode {
+    /// Socket addr of node.
+    saddr: SocketAddr,
+    /// Public Key of the node.
+    pub pk: PublicKey,
+}
+
+/// Size in bytes of serialized [`PackedNode`](./struct.PackedNode.html) with
+/// IPv4.
+pub const PACKED_NODE_IPV4_SIZE: usize = PUBLICKEYBYTES + 7;
+/// Size in bytes of serialized [`PackedNode`](./struct.PackedNode.html) with
+/// IPv6.
+pub const PACKED_NODE_IPV6_SIZE: usize = PUBLICKEYBYTES + 19;
+
+/** Serialize `PackedNode` into bytes.
+
+Can be either [`PACKED_NODE_IPV4_SIZE`]
+(./constant.PACKED_NODE_IPV4_SIZE.html) or [`PACKED_NODE_IPV6_SIZE`]
+(./constant.PACKED_NODE_IPV6_SIZE.html) bytes long, depending on whether
+IPv4 or IPv6 is being used.
+*/
+impl ToBytes for PackedNode {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_if_else!(self.saddr.is_ipv4(), gen_be_u8!(IpType::U4 as u8), gen_be_u8!(IpType::U6 as u8)) >>
+            gen_call!(|buf, addr| IpAddr::to_bytes(addr, buf), &self.saddr.ip()) >>
+            gen_be_u16!(self.saddr.port()) >>
+            gen_slice!(self.pk.as_ref())
+        )
+    }
+}
+
+/** Deserialize bytes into `PackedNode`. Returns `Error` if deseralizing
+failed.
+
+Can fail if:
+
+ - length is too short for given [`IpType`](./enum.IpType.html)
+ - PK can't be parsed
+
+Blindly trusts that provided `IpType` matches - i.e. if there are provided
+51 bytes (which is length of `PackedNode` that contains IPv6), and `IpType`
+says that it's actually IPv4, bytes will be parsed as if that was an IPv4
+address.
+*/
+named!(as_ipv4_packed_node<PackedNode>, do_parse!(
+    addr: call!(Ipv4Addr::from_bytes) >>
+    port: be_u16 >>
+    saddr: value!(SocketAddrV4::new(addr, port)) >>
+    pk: call!(PublicKey::from_bytes) >>
+    (PackedNode {
+        saddr: SocketAddr::V4(saddr),
+        pk: pk
+    })
+));
+
+// Parse bytes as an IPv6 PackedNode.
+named!(as_ipv6_packed_node<PackedNode>, do_parse!(
+    addr: call!(Ipv6Addr::from_bytes) >>
+    port: be_u16 >>
+    saddr: value!(SocketAddrV6::new(addr, port, 0, 0)) >>
+    pk: call!(PublicKey::from_bytes) >>
+    (PackedNode {
+        saddr: SocketAddr::V6(saddr),
+        pk: pk
+    })
+));
+
+impl FromBytes for PackedNode {
+    named!(from_bytes<PackedNode>, switch!(call!(IpType::from_bytes),
+        IpType::U4 => call!(as_ipv4_packed_node) |
```

Comment:
The code should look like:

```
    saddr: call(SocketAddr::from_bytes) >>
    pk: call!(PublicKey::from_bytes) >>
    (PackedNode {
        saddr: SocketAddr::V6(saddr),
        pk: pk
    })
```

The switch U4/U6 should be inside SocketAddr::from_bytes.